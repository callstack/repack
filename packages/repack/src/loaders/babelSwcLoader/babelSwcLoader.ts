import type { TransformOptions } from '@babel/core';
import type { LoaderContext, SwcLoaderOptions } from '@rspack/core';
import { transform } from '../babelLoader/babelLoader.js';
import type { BabelSwcLoaderOptions } from './options.js';
import {
  getSupportedSwcConfigurableTransforms,
  getSupportedSwcCustomTransforms,
  getSupportedSwcNormalTransforms,
} from './swc.js';
import {
  checkParallelModeAvailable,
  getExtraBabelPlugins,
  getProjectBabelConfig,
  getProjectRootPath,
  getSwcParserConfig,
  lazyGetSwc,
} from './utils.js';

type BabelTransform = [string, Record<string, any> | undefined];
type InputSourceMap = TransformOptions['inputSourceMap'];

export const raw = false;

function partitionTransforms(
  filename: string,
  babelTransforms: BabelTransform[]
) {
  let normalTransforms: string[] = [];
  let configurableTransforms: string[] = [];
  let customTransforms: string[] = [];

  let swcConfig: SwcLoaderOptions = {
    jsc: {
      parser: getSwcParserConfig(filename),
      transform: { react: { useBuiltins: true } },
    },
  };

  normalTransforms = getSupportedSwcNormalTransforms(babelTransforms);
  ({ swcConfig, transformNames: configurableTransforms } =
    getSupportedSwcConfigurableTransforms(babelTransforms, swcConfig));
  ({ swcConfig, transformNames: customTransforms } =
    getSupportedSwcCustomTransforms(babelTransforms, swcConfig));

  const includedSwcTransforms: string[] = [
    ...normalTransforms,
    ...configurableTransforms,
  ];

  const supportedSwcTransforms: string[] = [
    ...includedSwcTransforms,
    ...customTransforms,
  ];

  return { includedSwcTransforms, supportedSwcTransforms, swcConfig };
}

export default async function babelSwcLoader(
  this: LoaderContext<BabelSwcLoaderOptions>,
  source: string,
  sourceMap: string | undefined
) {
  this.cacheable();
  const callback = this.async();
  const loaderName = '@callstack/repack/babel-swc-loader';
  const logger = this.getLogger('BabelSwcLoader');
  const options = this.getOptions();

  if (!options.hideParallelModeWarning) {
    checkParallelModeAvailable(this, logger);
  }

  const inputSourceMap: InputSourceMap = sourceMap
    ? JSON.parse(sourceMap)
    : undefined;
  const lazyImports = options.lazyImports ?? true;
  const projectRoot = getProjectRootPath(this);

  const withSourceMaps = this.resourcePath.match(/node_modules/)
    ? false
    : this.sourceMap;

  const baseBabelConfig: TransformOptions = {
    caller: { name: loaderName },
    root: projectRoot,
    filename: this.resourcePath,
    sourceMaps: withSourceMaps,
    sourceFileName: this.resourcePath,
    sourceRoot: this.context,
    inputSourceMap: withSourceMaps ? inputSourceMap : undefined,
    ...options.babelOverrides,
  };

  try {
    const swc = await lazyGetSwc(this);
    // if swc is not available, use babel to transform everything
    if (!swc) {
      const { code, map } = await transform(source, baseBabelConfig, {
        hermesParserPath: options.hermesParserPath,
        hermesParserOverrides: options.hermesParserOverrides,
      });
      callback(null, code ?? undefined, map ?? undefined);
      return;
    }

    const babelConfig = getProjectBabelConfig(this.resourcePath, projectRoot);
    const detectedBabelTransforms =
      babelConfig.plugins?.map((p) => [p.key, p.options] as BabelTransform) ??
      [];

    const includeBabelPlugins = getExtraBabelPlugins(this.resourcePath);
    const { includedSwcTransforms, supportedSwcTransforms, swcConfig } =
      partitionTransforms(this.resourcePath, detectedBabelTransforms);

    const babelResult = await transform(source, baseBabelConfig, {
      excludePlugins: supportedSwcTransforms,
      includePlugins: includeBabelPlugins,
      hermesParserPath: options.hermesParserPath,
      hermesParserOverrides: options.hermesParserOverrides,
    });

    const finalSwcConfig: SwcLoaderOptions = {
      ...swcConfig,
      // set env based on babel transforms
      env: {
        // node supports everything and does not include
        // any transforms by default, so it can as a template
        targets: { node: 24 },
        include: includedSwcTransforms,
      },
      // set lazy imports based on loader options
      module: {
        ...swcConfig.module,
        lazy: lazyImports,
        type: swcConfig.module!.type,
      },
    };

    const swcResult = swc.transformSync(babelResult?.code!, {
      ...finalSwcConfig,
      caller: { name: loaderName },
      filename: this.resourcePath,
      configFile: false,
      swcrc: false,
      root: projectRoot ?? babelConfig.root ?? undefined,
      minify: false,
      sourceMaps: withSourceMaps,
      inputSourceMap: withSourceMaps
        ? JSON.stringify(babelResult?.map)
        : undefined,
      sourceFileName: this.resourcePath,
      sourceRoot: this.context!,
      ...options.swcOverrides,
    });

    callback(null, swcResult?.code, swcResult?.map);
  } catch (error) {
    callback(error as Error);
  }
}
