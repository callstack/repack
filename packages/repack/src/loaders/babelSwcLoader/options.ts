import type { TransformOptions } from '@babel/core';
import type { SwcLoaderOptions } from '@rspack/core';

type BabelOverrides = TransformOptions;
type SwcOverrides = Omit<SwcLoaderOptions, 'rspackExperiments'>;

export interface BabelSwcLoaderOptions {
  hideParallelModeWarning?: boolean;
  lazyImports?: boolean | string[];
  babelOverrides?: BabelOverrides;
  swcOverrides?: SwcOverrides;
}
