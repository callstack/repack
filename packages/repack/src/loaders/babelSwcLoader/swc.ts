import type { SwcLoaderOptions } from '@rspack/core';

const SWC_SUPPORTED_NORMAL_RULES = new Set([
  'transform-block-scoping',
  'transform-classes',
  'transform-class-static-block',
  'transform-destructuring',
  'transform-async-to-generator',
  'transform-async-generator-functions',
  'transform-unicode-regex',
  'transform-named-capturing-groups-regex',
  'transform-spread',
  'transform-parameters',
  'transform-function-name',
  'transform-logical-assignment-operators',
  'transform-sticky-regex',
  'transform-literals',
  'transform-optional-catch-binding',
  'transform-arrow-functions',
  'transform-numeric-separator',
  'transform-shorthand-properties',
  'transform-computed-properties',
]);

// NOTE: 'transform-class-properties' and 'transform-private-methods' are disabled here
// because, when combined with loose mode, these cause an internal swc error. Needs fixing upstream
const SWC_SUPPORTED_CONFIGURABLE_RULES = new Set([
  // 'transform-class-properties',
  // 'transform-private-methods',
  'transform-private-property-in-object',
  'transform-object-rest-spread',
  'transform-optional-chaining',
  'transform-nullish-coalescing-operator',
  'transform-for-of',
]);

const SWC_SUPPORTED_CUSTOM_RULES = new Set([
  'transform-runtime',
  'transform-react-jsx-self',
  'transform-react-jsx-source',
  'transform-react-jsx',
  'transform-modules-commonjs',
  'proposal-export-default-from',
]);

function getTransformRuntimeConfig(
  swcConfig: SwcLoaderOptions
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      externalHelpers: true,
    },
  };
}

function getTransformReactDevelopmentConfig(
  swcConfig: SwcLoaderOptions
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      transform: {
        ...swcConfig.jsc?.transform,
        react: {
          ...swcConfig.jsc?.transform?.react,
          development: true,
        },
      },
    },
  };
}

function getTransformReactRuntimeConfig(
  swcConfig: SwcLoaderOptions,
  reactRuntimeConfig: Record<string, any> = {
    runtime: 'automatic',
  }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      transform: {
        ...swcConfig.jsc?.transform,
        react: {
          ...swcConfig.jsc?.transform?.react,
          runtime: reactRuntimeConfig.runtime,
          importSource: reactRuntimeConfig.importSource,
        },
      },
    },
  };
}

function getTransformModulesCommonjsConfig(
  swcConfig: SwcLoaderOptions,
  moduleConfig: Record<string, any> = {
    strict: true,
    strictMode: true,
    allowTopLevelThis: true,
  }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    module: {
      ...swcConfig.module,
      type: 'commonjs',
      strict: Boolean(moduleConfig.strict),
      strictMode: Boolean(moduleConfig.strictMode),
      allowTopLevelThis: Boolean(moduleConfig.allowTopLevelThis),
      ignoreDynamic: true,
    },
  };
}

function getTransformExportDefaultFromConfig(
  swcConfig: SwcLoaderOptions
): SwcLoaderOptions {
  if (swcConfig.jsc?.parser?.syntax === 'typescript') {
    return swcConfig;
  }
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      parser: {
        ...swcConfig.jsc?.parser,
        syntax: 'ecmascript',
        exportDefaultFrom: true,
      },
    },
  };
}

function getTransformClassPropertiesConfig(
  swcConfig: SwcLoaderOptions,
  ruleConfig: Record<string, any> = { loose: false }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      assumptions: {
        ...swcConfig.jsc?.assumptions,
        setPublicClassFields:
          swcConfig.jsc?.assumptions?.setPublicClassFields || ruleConfig.loose,
      },
    },
  };
}

function getTransformPrivateMethodsPropertyConfig(
  swcConfig: SwcLoaderOptions,
  ruleConfig: Record<string, any> = { loose: false }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      assumptions: {
        ...swcConfig.jsc?.assumptions,
        privateFieldsAsProperties:
          swcConfig.jsc?.assumptions?.privateFieldsAsProperties ||
          ruleConfig.loose,
        setPublicClassFields:
          swcConfig.jsc?.assumptions?.setPublicClassFields || ruleConfig.loose,
      },
    },
  };
}

function getTransformObjectRestSpreadConfig(
  swcConfig: SwcLoaderOptions,
  ruleConfig: Record<string, any> = { loose: false }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      assumptions: {
        ...swcConfig.jsc?.assumptions,
        setSpreadProperties:
          swcConfig.jsc?.assumptions?.setSpreadProperties || ruleConfig.loose,
      },
    },
  };
}

function getTransformOptionalChainingNullishCoalescingConfig(
  swcConfig: SwcLoaderOptions,
  ruleConfig: Record<string, any> = { loose: false }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      assumptions: {
        ...swcConfig.jsc?.assumptions,
        noDocumentAll:
          swcConfig.jsc?.assumptions?.noDocumentAll || ruleConfig.loose,
      },
    },
  };
}

function getTransformForOfConfig(
  swcConfig: SwcLoaderOptions,
  ruleConfig: Record<string, any> = { loose: false }
): SwcLoaderOptions {
  return {
    ...swcConfig,
    jsc: {
      ...swcConfig.jsc,
      assumptions: {
        ...swcConfig.jsc?.assumptions,
        skipForOfIteratorClosing:
          swcConfig.jsc?.assumptions?.skipForOfIteratorClosing ||
          ruleConfig.loose,
      },
    },
  };
}

const SWC_SUPPORTED_CONFIGURABLE_RULES_MAP = {
  'transform-class-properties': getTransformClassPropertiesConfig,
  'transform-private-methods': getTransformPrivateMethodsPropertyConfig,
  'transform-private-property-in-object':
    getTransformPrivateMethodsPropertyConfig,
  'transform-object-rest-spread': getTransformObjectRestSpreadConfig,
  'transform-optional-chaining':
    getTransformOptionalChainingNullishCoalescingConfig,
  'transform-nullish-coalescing-operator':
    getTransformOptionalChainingNullishCoalescingConfig,
  'transform-for-of': getTransformForOfConfig,
};

const SWC_SUPPORTED_CUSTOM_RULES_MAP = {
  'transform-runtime': getTransformRuntimeConfig,
  'transform-react-jsx': getTransformReactRuntimeConfig,
  'transform-react-jsx-self': getTransformReactDevelopmentConfig,
  'transform-react-jsx-source': getTransformReactDevelopmentConfig,
  'transform-modules-commonjs': getTransformModulesCommonjsConfig,
  'proposal-export-default-from': getTransformExportDefaultFromConfig,
};

export function getSupportedSwcNormalTransforms(
  transforms: [string, Record<string, any> | undefined][]
) {
  return transforms
    .filter(([transform]) => SWC_SUPPORTED_NORMAL_RULES.has(transform))
    .map(([transform]) => transform);
}

export function getSupportedSwcConfigurableTransforms(
  transforms: [string, Record<string, any> | undefined][],
  swcConfig: SwcLoaderOptions
) {
  const transformNames = transforms
    .filter(([transform]) => SWC_SUPPORTED_CONFIGURABLE_RULES.has(transform))
    .map(([transform]) => transform);
  const finalSwcConfig = transforms
    .filter(([transform]) => SWC_SUPPORTED_CONFIGURABLE_RULES.has(transform))
    .reduce((config, [transform, transformConfig]) => {
      const handler =
        SWC_SUPPORTED_CONFIGURABLE_RULES_MAP[
          transform as keyof typeof SWC_SUPPORTED_CONFIGURABLE_RULES_MAP
        ];
      return handler(config, transformConfig);
    }, swcConfig);
  return { swcConfig: finalSwcConfig, transformNames };
}

export function getSupportedSwcCustomTransforms(
  transforms: [string, Record<string, any> | undefined][],
  swcConfig: SwcLoaderOptions
) {
  const transformNames = transforms
    .filter(([transform]) => SWC_SUPPORTED_CUSTOM_RULES.has(transform))
    .map(([transform]) => transform);
  const finalSwcConfig = transforms
    .filter(([transform]) => SWC_SUPPORTED_CUSTOM_RULES.has(transform))
    .reduce((config, [transform, transformConfig]) => {
      const handler =
        SWC_SUPPORTED_CUSTOM_RULES_MAP[
          transform as keyof typeof SWC_SUPPORTED_CUSTOM_RULES_MAP
        ];
      return handler(config, transformConfig);
    }, swcConfig);
  return { swcConfig: finalSwcConfig, transformNames };
}
