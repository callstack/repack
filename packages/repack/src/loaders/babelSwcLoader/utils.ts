import { loadOptions } from '@babel/core';
import type {
  LoaderContext,
  SwcLoaderParserConfig,
  experiments,
} from '@rspack/core';

type Swc = (typeof experiments)['swc'];
type Logger = ReturnType<LoaderContext['getLogger']>;

export function isTypeScriptSource(fileName: string) {
  return !!fileName && fileName.endsWith('.ts');
}

export function isTSXSource(fileName: string) {
  return !!fileName && fileName.endsWith('.tsx');
}

export function getProjectBabelConfig(filename: string, projectRoot?: string) {
  const babelConfig = loadOptions({ filename, root: projectRoot });
  return babelConfig ?? {};
}

export function getExtraBabelPlugins(filename: string) {
  const extraBabelPlugins: Array<string | [string, Record<string, any>]> = [];
  // add TS syntax plugins since RN preset
  // only uses transform-typescript plugin
  // which includes the syntax-typescript plugin
  if (isTypeScriptSource(filename)) {
    extraBabelPlugins.push([
      '@babel/plugin-syntax-typescript',
      { isTSX: false, allowNamespaces: true },
    ]);
  } else if (isTSXSource(filename)) {
    extraBabelPlugins.push([
      '@babel/plugin-syntax-typescript',
      { isTSX: true, allowNamespaces: true },
    ]);
  }
  return extraBabelPlugins;
}

export function getSwcParserConfig(filename: string): SwcLoaderParserConfig {
  if (isTypeScriptSource(filename)) {
    return { syntax: 'typescript', tsx: false };
  }
  if (isTSXSource(filename)) {
    return { syntax: 'typescript', tsx: true };
  }
  // include JSX in .js files
  return { syntax: 'ecmascript', jsx: true };
}

function isWebpackBackend(loaderContext: LoaderContext) {
  // parallel loader in Rspack has mocked compiler object without most props
  // but in non-parallel mode, full compiler object is available so we can do the proper check
  // we distinguish between parallel and non-parallel mode by checking if the compiler has a version property
  // and then proceed to check if it's a Rspack compiler the official way
  if (
    'webpack' in loaderContext._compiler &&
    'version' in loaderContext._compiler.webpack &&
    loaderContext._compiler.webpack.version
  ) {
    return !('rspackVersion' in loaderContext._compiler.webpack);
  }
  // in threaded-loader mode, _compiler.webpack is undefined
  // loaderContext.webpack exists in webpack but not in Rspack
  // in case it's added in the future, it should be followed by similar rspack prop
  if ('webpack' in loaderContext && loaderContext.webpack) {
    return !('rspack' in loaderContext && loaderContext.rspack);
  }
  // if both checks fail, we assume it's a Rspack compiler
  return false;
}

const disabledParalleModeWarning = [
  'You have enabled `experiments.parallelLoader` but forgot to enable it for this loader.',
  'To enable parallel mode for this loader you need to add `parallel: true` to the loader rule.',
  'See how to do it in the official Rspack docs:',
  'https://rspack.rs/config/experiments#experimentsparallelloader.',
  'If this is intentional, you can disable this warning',
  'by setting `hideParallelModeWarning` in the loader options.',
].join(' ');

let parallelModeWarningDisplayed = false;

export function checkParallelModeAvailable(
  loaderContext: LoaderContext,
  logger: Logger
) {
  // only Rspack supports parallel mode
  if (parallelModeWarningDisplayed || isWebpackBackend(loaderContext)) {
    return;
  }
  // in parallel mode compiler.options.experiments are not available
  // but since we're already running in parallel mode, we can ignore this check
  if (loaderContext._compiler.options?.experiments?.parallelLoader) {
    parallelModeWarningDisplayed = true;
    logger.warn(disabledParalleModeWarning);
  }
}

export function getProjectRootPath(
  loaderContext: LoaderContext
): string | undefined {
  // parallel loaders in Rspack had a bug where rootContext
  // was the same as context, so we check if they are different
  if (loaderContext.rootContext !== loaderContext.context) {
    return loaderContext.rootContext;
  }
  return undefined;
}

function safelyResolve(path: string, from: string): string | null {
  try {
    return require.resolve(path, { paths: [from] });
  } catch {
    return null;
  }
}

async function getSwcModule(loaderContext: LoaderContext): Promise<Swc | null> {
  const projectRoot = getProjectRootPath(loaderContext) ?? process.cwd();
  const isWebpack = isWebpackBackend(loaderContext);
  if (!isWebpack) {
    // happy path - rspack & exposed swc
    // use optional chaining to avoid type errors when using parallel loader
    if (loaderContext._compiler.rspack?.experiments?.swc) {
      return loaderContext._compiler.rspack.experiments.swc;
    }
    // fallback to checking for `@rspack/core` installed in the project
    // use optional chaining to avoid type errors when there is no experiments.swc
    const rspackCorePath = safelyResolve('@rspack/core', projectRoot);
    if (rspackCorePath && !isWebpack) {
      const rspack = await import(rspackCorePath);
      if ('default' in rspack) {
        return rspack.default?.experiments?.swc ?? null;
      }
      if (rspack) {
        return rspack?.experiments?.swc ?? null;
      }
    }
  }
  // fallback to checking for `@swc/core` installed in the project
  // this can be in both webpack & rspack projects
  const swcCorePath = safelyResolve('@swc/core', projectRoot);
  if (swcCorePath) {
    const swc = await import(swcCorePath);
    if ('default' in swc) {
      return swc.default as Swc;
    }
    if (swc) {
      return swc as Swc;
    }
  }
  // at this point, we've tried all possible ways to get swc and failed
  return null;
}

function createLazyGetSwc(): (
  loaderContext: LoaderContext
) => Promise<Swc | null> {
  let swc: Swc | Promise<Swc | null> | null | undefined;

  const getSwc = async (loaderContext: LoaderContext) => {
    if (swc === undefined) {
      swc = getSwcModule(loaderContext);
    }
    return await swc;
  };
  return getSwc;
}

export const lazyGetSwc = createLazyGetSwc();
