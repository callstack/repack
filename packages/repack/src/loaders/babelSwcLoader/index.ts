import babelSwcLoader, { raw } from './babelSwcLoader.js';

export { raw };
export default babelSwcLoader;
