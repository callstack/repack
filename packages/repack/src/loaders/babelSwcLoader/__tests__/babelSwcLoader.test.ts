import { partitionTransforms } from '../babelSwcLoader.js';

type TransformEntry = [string, Record<string, any> | undefined];

describe('partitionTransforms', () => {
  it('partitions normal, configurable, and custom transforms preserving order', () => {
    const transforms: TransformEntry[] = [
      ['transform-block-scoping', {}],
      ['transform-react-jsx', {}],
      ['transform-object-rest-spread', { loose: true }],
      ['transform-class-properties', { loose: true }], // disabled configurable
      ['transform-private-methods', { loose: true }], // disabled configurable
      ['unknown-plugin', {}],
      ['transform-modules-commonjs', {}],
      ['transform-classes', {}],
      ['transform-for-of', {}],
    ];

    const { includedSwcTransforms, supportedSwcTransforms } =
      partitionTransforms('/virtual/file.tsx', transforms);

    expect(includedSwcTransforms).toEqual([
      'transform-block-scoping',
      'transform-classes',
      'transform-object-rest-spread',
      'transform-for-of',
    ]);

    expect(supportedSwcTransforms).toEqual([
      'transform-block-scoping',
      'transform-classes',
      'transform-object-rest-spread',
      'transform-for-of',
      'transform-react-jsx',
      'transform-modules-commonjs',
    ]);
  });

  it('ignores unsupported and disabled transforms (snapshot)', () => {
    const transforms: TransformEntry[] = [
      ['transform-class-properties', { loose: true }], // disabled
      ['transform-private-methods', { loose: true }], // disabled
      ['unknown-plugin', {}],
    ];

    const result = partitionTransforms('/virtual/file.js', transforms);

    expect(result).toMatchSnapshot('unsupported and disabled');
  });

  it('only custom transforms are excluded from included set but present in supported set (snapshot)', () => {
    const transforms: TransformEntry[] = [
      ['transform-react-jsx', {}],
      ['transform-modules-commonjs', {}],
      ['proposal-export-default-from', {}],
    ];

    const result = partitionTransforms('/virtual/file.jsx', transforms);

    expect(result).toMatchSnapshot('custom-only supported set');
  });

  it('returns empty arrays when no transforms are specified (snapshot)', () => {
    const result = partitionTransforms('/virtual/empty.ts', []);

    expect(result).toMatchSnapshot('empty arrays');
  });
});
