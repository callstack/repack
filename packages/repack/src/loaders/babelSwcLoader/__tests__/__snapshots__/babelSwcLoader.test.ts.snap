// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`partitionTransforms ignores unsupported and disabled transforms (snapshot): unsupported and disabled 1`] = `
{
  "includedSwcTransforms": [],
  "supportedSwcTransforms": [],
  "swcConfig": {
    "jsc": {
      "parser": {
        "jsx": true,
        "syntax": "ecmascript",
      },
      "transform": {
        "react": {
          "useBuiltins": true,
        },
      },
    },
  },
}
`;

exports[`partitionTransforms only custom transforms are excluded from included set but present in supported set (snapshot): custom-only supported set 1`] = `
{
  "includedSwcTransforms": [],
  "supportedSwcTransforms": [
    "transform-react-jsx",
    "transform-modules-commonjs",
    "proposal-export-default-from",
  ],
  "swcConfig": {
    "jsc": {
      "parser": {
        "exportDefaultFrom": true,
        "jsx": true,
        "syntax": "ecmascript",
      },
      "transform": {
        "react": {
          "importSource": undefined,
          "runtime": undefined,
          "useBuiltins": true,
        },
      },
    },
    "module": {
      "allowTopLevelThis": false,
      "ignoreDynamic": true,
      "strict": false,
      "strictMode": false,
      "type": "commonjs",
    },
  },
}
`;

exports[`partitionTransforms returns empty arrays when no transforms are specified (snapshot): empty arrays 1`] = `
{
  "includedSwcTransforms": [],
  "supportedSwcTransforms": [],
  "swcConfig": {
    "jsc": {
      "parser": {
        "syntax": "typescript",
        "tsx": false,
      },
      "transform": {
        "react": {
          "useBuiltins": true,
        },
      },
    },
  },
}
`;
